reps/2022-08-31-actor-affinity-apis.md

```diff
@@ -0,0 +1,405 @@
+```python
+SchedulingStrategyT = Union[None, str,
+                            PlacementGroupSchedulingStrategy,
+                            NodeAffinitySchedulingStrategy,
+                            ActorAffinitySchedulingStrategy]
+
+class ActorAffinitySchedulingStrategy:
+    def __init__(self, match_expressions: List[ActorAffinityMatchExpression]):
+        self.match_expressions = match_expressions
+
+class ActorAffinityMatchExpression:
+    """An expression used to represent an actor's affinity.
+    Attributes:
+        key: the key of label
+        operator: IN、NOT_IN、EXISTS、DOES_NOT_EXIST,
+                  if EXISTS、DOES_NOT_EXIST, values set []
+        values: a list of label value
+        soft: ...
+    """
+    def __init__(self, key: str, operator: ActorAffinityOperator,
+                 values: List[str], soft: bool):
+        self.key = key
+        self.operator = operator
+        self.values = values
+        self.soft = soft
+
+```
+
+Python API example:  
+Step 1: Set the labels for this actor.   
+Each label consists of a key and value of type string.
+```python
+actor_1 = Actor.options(labels={
+    "location": "dc_1"
+}).remote()
+```
+
+Step 2: Set actor affinity strategy.  
+1. The target actor is expected to be scheduled with the actors whose label key is "location" and value in ["dc-1"].
+```python
+match_expressions = [
+    ActorAffinityMatchExpression("location", ActorAffinityOperator.IN, ["dc_1"], False)
+]
+actor_affinity_strategy = ActorAffinitySchedulingStrategy(match_expressions)
+actor = Actor.options(scheduling_strategy = actor_affinity_strategy).remote()
+```
+
+2. The target actor is not expected to be scheduled
+with the actors whose label key is "location" and value in ["dc-1"].
+```python
+match_expressions = [
+ ActorAffinityMatchExpression("location", ActorAffinityOperator.NOT_IN, ["dc_1"], False)
+]
+actor_affinity_strategy = ActorAffinitySchedulingStrategy(match_expressions)
+actor = Actor.options(scheduling_strategy = actor_affinity_strategy).remote()
+```
+
+3. The target actor is expected to be scheduled with the actors whose label key exists "location".
+```python
+match_expressions = [
+ ActorAffinityMatchExpression("location", ActorAffinityOperator.EXISTS, [], False)
+]
+actor_affinity_strategy = ActorAffinitySchedulingStrategy(match_expressions)
+actor = Actor.options(scheduling_strategy = actor_affinity_strategy).remote()
+```
+
+4. The target actor is not expected to be scheduled with the actors whose label key exists "location".
+```python
+match_expressions = [
+ ActorAffinityMatchExpression("location", ActorAffinityOperator.DOES_NOT_EXIST, [], False)
+]
+actor_affinity_strategy = ActorAffinitySchedulingStrategy(match_expressions)
+actor = Actor.options(scheduling_strategy = actor_affinity_strategy).remote()
+```
+
+5. You can also set multiple expressions at the same time, and multiple expressions need to be satisfied when scheduling.
+```python
+match_expressions = [
```

Comment:
Don't support OR semantic.  Because of the AND semantics +  the operation in "IN/ NOT_IN / EXISTS/ DOES_NOT_EXIST", all scenarios can already be expressed. Adding the semantics of OR will increase the complexity of the API, making it difficult for users to get started.